tools/istio-iptables/pkg/dependencies/implementation.go

```diff
@@ -99,16 +102,147 @@ func (v IptablesVersion) IsWriteCmd(cmd constants.IptablesCmd) bool {
 const (
-	iptablesBin         = "iptables"
-	iptablesNftBin      = "iptables-nft"
-	iptablesLegacyBin   = "iptables-legacy"
-	ip6tablesBin        = "ip6tables"
-	ip6tablesNftBin     = "ip6tables-nft"
-	ip6tablesLegacyBin  = "ip6tables-legacy"
-	iptablesRestoreBin  = "iptables-restore"
-	ip6tablesRestoreBin = "ip6tables-restore"
+	iptablesBin              = "iptables"
+	iptablesNftBin           = "iptables-nft"
+	iptablesLegacyBin        = "iptables-legacy"
+	ip6tablesBin             = "ip6tables"
+	ip6tablesNftBin          = "ip6tables-nft"
+	ip6tablesLegacyBin       = "ip6tables-legacy"
+	iptablesRestoreBin       = "iptables-restore"
+	ip6tablesRestoreBin      = "ip6tables-restore"
+	ModulesFile              = "/proc/modules"
+	iptablesLegacyModuleName = "ip_tables"
+	iptablesNftModuleName    = "nf_tables"
 )
 
+func NewIpTablesVersion(bin string, versionOutput string) (*IptablesVersion, error) {
+	var parsedVer *utilversion.Version
+	if !strings.Contains(versionOutput, "unrecognized option") {
+		var parseErr error
+		parsedVer, parseErr = parseIptablesVer(versionOutput)
+		if parseErr != nil {
+			return nil, fmt.Errorf("iptables version %q is not a valid version string: %v", versionOutput, parseErr)
+		}
+	} else {
+		log.Warnf("found iptables binary %s, but it does not appear to support the '--version' flag, assuming very old legacy version", bin)
+		parsedVer = utilversion.MustParseGeneric("0.0.0")
+	}
+	version := &IptablesVersion{
+		DetectedBinary:        bin,
+		DetectedSaveBinary:    fmt.Sprintf("%s-save", bin),
+		DetectedRestoreBinary: fmt.Sprintf("%s-restore", bin),
+		Version:               parsedVer,
+	}
+	rulesDump, err := exec.Command(version.DetectedSaveBinary).CombinedOutput()
+	if err != nil {
+		return nil, err
+	}
+	if strings.Count(string(rulesDump), "\n") >= 3 {
+		version.ExistingRules = true
+	} else {
+		version.ExistingRules = false
+	}
+	return version, nil
+}
+
+func LoadProcModules() (map[string]struct{}, error) {
+	file, err := os.Open(ModulesFile)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	modules := make(map[string]struct{})
+	scanner := bufio.NewScanner(file)
+	for scanner.Scan() {
+		line := scanner.Text()
+		fields := strings.Fields(line)
+		if len(fields) > 0 {
+			moduleName := fields[0]
+			modules[moduleName] = struct{}{}
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+	return modules, nil
+}
+
+func getIptablesNFTBin(isIpV6 bool) string {
+	if isIpV6 {
+		return ip6tablesNftBin
+	} else {
+		return iptablesNftBin
+	}
+}
+
+func getIptablesLegacyBin(isIpV6 bool) string {
+	if isIpV6 {
+		return ip6tablesLegacyBin
+	} else {
+		return iptablesLegacyBin
+	}
+}
+
+func getIptablesBin(isIpV6 bool) string {
+	if isIpV6 {
+		return ip6tablesBin
+	} else {
+		return iptablesBin
+	}
+}
+
+func GetNFTVersion(modules map[string]struct{}, ipV6 bool) (*IptablesVersion, error) {
+	if modules == nil {
+		return nil, fmt.Errorf("nil modules map")
+	}
```

Comment:
nit: should this say something along the lines of
```suggestion
	if modules == nil {
		return nil, fmt.Errorf("unable to get nft version because modules map is nil")
	}
```

Reply:
i have modified the logic to ensure LoadProcModules won't return a nil map when there is no error.
so we can delete this 'if' here